diff --git a/src/plugin.c b/src/plugin.c
index 7cd7320..e7e79c6 100644
--- a/src/plugin.c
+++ b/src/plugin.c
@@ -906,3 +906,9 @@ void popup_at_button (LXPanel *panel, GtkWidget *window, GtkWidget *button, gpoi
     g_signal_connect (G_OBJECT (window), "map-event", G_CALLBACK (popup_mapped), plugin);
     g_signal_connect (G_OBJECT (window), "button-press-event", G_CALLBACK (popup_button_press), plugin);
 }
+
+gboolean is_pi (void)
+{
+    if (!access ("/boot/firmware/config.txt", R_OK)) return TRUE;
+    return FALSE;
+}
diff --git a/src/plugin.h b/src/plugin.h
index f752218..2facc01 100644
--- a/src/plugin.h
+++ b/src/plugin.h
@@ -417,6 +417,8 @@ extern void graph_free (PluginGraph *graph);
 
 extern void popup_at_button (LXPanel *panel, GtkWidget *window, GtkWidget *button, gpointer plugin);
 
+extern gboolean is_pi (void);
+
 #define wrap_new_menu_item(plugin,text,maxlen,icon) lxpanel_plugin_new_menu_item(plugin->panel,text,maxlen,icon)
 #define wrap_set_menu_icon(plugin,image,icon) lxpanel_plugin_set_menu_icon(plugin->panel,image,icon)
 #define wrap_set_taskbar_icon(plugin,image,icon) lxpanel_plugin_set_taskbar_icon(plugin->panel,image,icon)
