diff --git a/plugins/launchtaskbar.c b/plugins/launchtaskbar.c
index 969ef21..7981d18 100644
--- a/plugins/launchtaskbar.c
+++ b/plugins/launchtaskbar.c
@@ -1862,12 +1862,14 @@ static void launchtaskbar_panel_configuration_changed(LXPanel *panel, GtkWidget
 
     if (ltbp->lb_built)
     {
-        GList *l;
-        for (l = gtk_container_get_children(GTK_CONTAINER(ltbp->lb_icon_grid)); l != NULL; l = l->next)
+        GList *list, *l;
+        list = gtk_container_get_children(GTK_CONTAINER(ltbp->lb_icon_grid));
+        for (l = list; l != NULL; l = l->next)
         {
             LaunchButton *btn = (LaunchButton *) l->data;
             launch_button_update_icon (btn, new_icon_size - ICON_BUTTON_TRIM);
         }
+        g_list_free (list);
         panel_icon_grid_set_geometry(PANEL_ICON_GRID(ltbp->lb_icon_grid),
                                      panel_get_orientation(panel),
                                      new_icon_size, new_icon_size,
diff --git a/plugins/menu.c b/plugins/menu.c
index 8166189..0c20f30 100644
--- a/plugins/menu.c
+++ b/plugins/menu.c
@@ -180,6 +180,7 @@ static void on_menu_item_map(GtkWidget *mi, menup *m)
 #if GTK_CHECK_VERSION(3, 0, 0)
   GList *children = gtk_container_get_children (GTK_CONTAINER (gtk_bin_get_child (GTK_BIN (mi))));
   GtkImage *img = children->data;  // this should always be the image...
+  g_list_free (children);
 #else
   GtkImage* img = GTK_IMAGE(gtk_image_menu_item_get_image(GTK_IMAGE_MENU_ITEM(mi)));
 #endif
@@ -572,13 +573,14 @@ static void _unload_old_icons(GtkMenu* menu, GtkIconTheme* theme, menup* m)
                 GtkWidget *box = gtk_bin_get_child (GTK_BIN (item));
                 if (GTK_IS_BOX (box))
                 {
-                    GList *children = gtk_container_get_children (GTK_CONTAINER (box));
-                    if (GTK_IS_IMAGE (children->data))
+                    GList *bchildren = gtk_container_get_children (GTK_CONTAINER (box));
+                    if (GTK_IS_IMAGE (bchildren->data))
                     {
-                        img = GTK_IMAGE (children->data);
+                        img = GTK_IMAGE (bchildren->data);
                         gtk_image_clear(img);
                         if (gtk_widget_get_mapped (GTK_WIDGET(img))) on_menu_item_map(GTK_WIDGET(item), m);
                     }
+                    g_list_free (bchildren);
                 }
             }
 #else
