diff --git a/plugins/menu.c b/plugins/menu.c
index 781ad26..15306ec 100644
--- a/plugins/menu.c
+++ b/plugins/menu.c
@@ -688,7 +688,7 @@ reload_system_menu( menup* m, GtkMenu* menu )
 static void show_menu( GtkWidget* widget, menup* m, int btn, guint32 time )
 {
 #if GTK_CHECK_VERSION(3, 0, 0)
-    gtk_menu_popup_at_widget (GTK_MENU(m->menu), widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, NULL);
+    gtk_menu_popup_at_widget (GTK_MENU(m->menu), widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, NULL);
 #else
     gtk_menu_popup(GTK_MENU(m->menu),
                    NULL, NULL,
diff --git a/plugins/netstat/netstat.c b/plugins/netstat/netstat.c
index 15894e9..c87f985 100644
--- a/plugins/netstat/netstat.c
+++ b/plugins/netstat/netstat.c
@@ -251,7 +251,7 @@ static gint menupopup(GtkWidget *widget, GdkEvent *event, netdev_info *ni)
             /* wireless device */
             if (ni->netdev_list->info.wireless) {
 #if GTK_CHECK_VERSION(3, 0, 0)
-                gtk_menu_popup_at_widget (GTK_MENU(wireless_menu(ni)), widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, event);
+                gtk_menu_popup_at_widget (GTK_MENU(wireless_menu(ni)), widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, event);
 #else
                 gtk_menu_popup(GTK_MENU(wireless_menu(ni)), NULL, NULL, NULL, NULL, event_button->button, event_button->time);
 #endif
@@ -278,7 +278,7 @@ static gint menupopup(GtkWidget *widget, GdkEvent *event, netdev_info *ni)
             gtk_widget_show_all(menu);
 
 #if GTK_CHECK_VERSION(3, 0, 0)
-            gtk_menu_popup_at_widget (GTK_MENU(menu), widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, event);
+            gtk_menu_popup_at_widget (GTK_MENU(menu), widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, event);
 #else
             gtk_menu_popup(GTK_MENU(menu), NULL, NULL, NULL, NULL, event_button->button, event_button->time);
 #endif
diff --git a/plugins/task-button.c b/plugins/task-button.c
index b5eae3f..691462c 100644
--- a/plugins/task-button.c
+++ b/plugins/task-button.c
@@ -1296,7 +1296,7 @@ static gboolean task_button_button_press_event(GtkWidget *widget, GdkEventButton
         /* attach menu to the widget and show it */
         gtk_menu_attach_to_widget(GTK_MENU(menu), widget, NULL);
 #if GTK_CHECK_VERSION(3, 0, 0)
-        gtk_menu_popup_at_widget (GTK_MENU(menu), widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
+        gtk_menu_popup_at_widget (GTK_MENU(menu), widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
 #else
         gtk_menu_popup(GTK_MENU(menu), NULL, NULL, taskbar_popup_set_position,
                        tb, event->button, event->time);
@@ -1382,7 +1382,7 @@ static gboolean task_button_button_release_event(GtkWidget *widget, GdkEventButt
             gtk_widget_show_all(GTK_WIDGET(tb->menu_list));
             gtk_menu_attach_to_widget(tb->menu_list, widget, NULL);
 #if GTK_CHECK_VERSION(3, 0, 0)
-            gtk_menu_popup_at_widget (tb->menu_list, widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
+            gtk_menu_popup_at_widget (tb->menu_list, widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
 #else
             gtk_menu_popup(tb->menu_list, NULL, NULL, taskbar_popup_set_position,
                            tb, event->button, event->time);
diff --git a/src/panel.c b/src/panel.c
index 6e0064d..3faa952 100644
--- a/src/panel.c
+++ b/src/panel.c
@@ -510,7 +510,7 @@ static gboolean lxpanel_button_press(GtkWidget *widget, GdkEventButton *event)
     {
         GtkMenu* popup = (GtkMenu*) lxpanel_get_plugin_menu(panel, NULL, FALSE);
 #if GTK_CHECK_VERSION(3, 0, 0)
-        gtk_menu_popup_at_widget (popup, widget, GDK_GRAVITY_NORTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
+        gtk_menu_popup_at_widget (popup, widget, GDK_GRAVITY_SOUTH_WEST, GDK_GRAVITY_NORTH_WEST, (GdkEvent *) event);
 #else
         gtk_menu_popup(popup, NULL, NULL, NULL, NULL, event->button, event->time);
 #endif
