diff --git a/src/notify.c b/src/notify.c
index b61e7cb..0727338 100644
--- a/src/notify.c
+++ b/src/notify.c
@@ -64,6 +64,33 @@ static gboolean window_click (GtkWidget *widget, GdkEventButton *event, NotifyWi
 /* Private functions */
 /*----------------------------------------------------------------------------*/
 
+/* Calculate position; based on xpanel_plugin_popup_set_position_helper */
+
+static void notify_position_helper (LXPanel *p, GtkWidget *popup, gint *px, gint *py)
+{
+    GdkMonitor *monitor;
+    GdkRectangle pop_geom, mon_geom, pan_geom;
+
+    /* Get the geometry of the monitor on which the panel is displayed */
+    monitor = gdk_display_get_monitor_at_window (gtk_widget_get_display (p->priv->box), gtk_widget_get_window (p->priv->box));
+    gdk_monitor_get_geometry (monitor, &mon_geom);
+
+    /* Get the geometry of the panel */
+    gdk_window_get_frame_extents (gtk_widget_get_window (p->priv->box), &pan_geom);
+
+    /* Get the geometry of the popup menu */
+    gtk_widget_realize (popup);
+    gdk_window_get_frame_extents (gtk_widget_get_window (popup), &pop_geom);
+
+    /* By default, notifications go in the top right corner of the monitor with the panel */
+    *px = mon_geom.x + mon_geom.width - pop_geom.width;
+    *py = mon_geom.y;
+
+    /* Shift if panel is in the way...*/
+    if (p->priv->edge == EDGE_TOP) *py += pan_geom.height;
+    if (p->priv->edge == EDGE_RIGHT) *px -= pan_geom.width;
+}
+
 /* Create a notification window and position appropriately */
 
 static void show_message (LXPanel *panel, NotifyWindow *nw, char *str)
@@ -112,11 +139,8 @@ static void show_message (LXPanel *panel, NotifyWindow *nw, char *str)
     gtk_widget_show_all (nw->popup);
     gtk_widget_hide (nw->popup);
 
-    plugins = gtk_container_get_children (GTK_CONTAINER (panel->priv->box));
-    lxpanel_plugin_popup_set_position_helper (panel, (GtkWidget *) (g_list_last (plugins))->data, nw->popup, &x, &y);
-    if (panel->priv->edge == EDGE_BOTTOM) gdk_window_move (gtk_widget_get_window (nw->popup), x, SPACING);
-    else gdk_window_move (gtk_widget_get_window (nw->popup), x, y);
-    g_list_free (plugins);
+    notify_position_helper (panel, nw->popup, &x, &y);
+    gdk_window_move (gtk_widget_get_window (nw->popup), x, y);
 
     gdk_window_set_events (gtk_widget_get_window (nw->popup), gdk_window_get_events (gtk_widget_get_window (nw->popup)) | GDK_BUTTON_PRESS_MASK);
     g_signal_connect (G_OBJECT (nw->popup), "button-press-event", G_CALLBACK (window_click), nw);
