diff --git a/src/main.c b/src/main.c
index 820e107..26c7b6c 100644
--- a/src/main.c
+++ b/src/main.c
@@ -658,7 +658,7 @@ static gboolean start_all_panels( )
     const gchar * const * dir;
 
     /* try user panels */
-    panel_dir = _user_config_file_name("panels", NULL);
+    panel_dir = _user_config_file_name(is_wizard () ? "wizard" : "panels", NULL);
 
     /* check to see if there are any panels which will display on monitor 0 */
     char *cmd = g_strdup_printf ("grep -l Global $(grep -L monitor=[1-9] %s/*) /dev/null | grep -c . | grep -q 0", panel_dir);
@@ -675,7 +675,7 @@ static gboolean start_all_panels( )
     dir = g_get_system_config_dirs();
     if (dir) while (dir[0])
     {
-        panel_dir = _system_config_file_name(dir[0], "panels");
+        panel_dir = _system_config_file_name(dir[0], is_wizard () ? "wizard" : "panels");
         _start_panels_from_dir(panel_dir, 0);
         g_free(panel_dir);
         if (all_panels != NULL)
diff --git a/src/misc.c b/src/misc.c
index ca38979..5fbf03a 100644
--- a/src/misc.c
+++ b/src/misc.c
@@ -1780,4 +1780,10 @@ int screen_height (GdkScreen *scr)
 }
 #endif
 
+gboolean is_wizard (void)
+{
+    if (!g_strcmp0 (getenv ("USER"), "rpi-first-boot-wizard")) return TRUE;
+    return FALSE;
+}
+
 /* vim: set sw=4 et sts=4 : */
diff --git a/src/misc.h b/src/misc.h
index 3ba9341..a4d87e2 100644
--- a/src/misc.h
+++ b/src/misc.h
@@ -289,6 +289,8 @@ extern int screen_width (GdkScreen *scr);
 extern int screen_height (GdkScreen *scr);
 #endif
 
+extern gboolean is_wizard (void);
+
 G_END_DECLS
 
 #endif
diff --git a/src/panel.c b/src/panel.c
index 7f8a604..ca218f3 100644
--- a/src/panel.c
+++ b/src/panel.c
@@ -457,10 +457,6 @@ static void lxpanel_size_allocate(GtkWidget *widget, GtkAllocation *a)
     p->ax = rect.x;
     p->ay = rect.y;
 
-#if GTK_CHECK_VERSION(3, 0, 0)
-    if (a->width != p->aw || a->height != p->ah) gtk_widget_set_size_request (widget, p->aw, p->ah);
-#endif
-
     if (rect.width != p->aw || rect.height != p->ah || x != p->ax || y != p->ay)
     {
         p->aw = rect.width;
@@ -508,6 +504,8 @@ static gboolean lxpanel_button_press(GtkWidget *widget, GdkEventButton *event)
         /* ignore clicks with modifiers */
         return FALSE;
 
+    if (is_wizard ()) return TRUE;
+
     if (event->button == 3) /* right button */
     {
         GtkMenu* popup = (GtkMenu*) lxpanel_get_plugin_menu(panel, NULL, FALSE);
diff --git a/src/plugin.c b/src/plugin.c
index effb97e..71c08f1 100644
--- a/src/plugin.c
+++ b/src/plugin.c
@@ -237,6 +237,7 @@ static gboolean lxpanel_plugin_button_press_event(GtkWidget *plugin, GdkEventBut
     if (event->button == 3 && /* right button */
         (event->state & gtk_accelerator_get_default_mod_mask()) == 0) /* no key */
     {
+        if (is_wizard ()) return TRUE;
 #ifdef ENABLE_NLS
 		// this message comes via the plugin, which will have switched to its own text domain, so 
 		// we need to switch back here...
