diff --git a/plugins/xkb/xkb-plugin.c b/plugins/xkb/xkb-plugin.c
index 6315dac..04f5850 100644
--- a/plugins/xkb/xkb-plugin.c
+++ b/plugins/xkb/xkb-plugin.c
@@ -711,8 +711,10 @@ static gboolean  change_opt_tree_model_foreach(GtkTreeModel *p_model,
         if(strlen(p_xkb->p_gstring_change_opt_partial->str))
         {
             g_string_append_c(p_xkb->p_gstring_change_opt_partial, ',');
+            g_string_append_c (p_xkb->p_gset_opts, ',');
         }
         g_string_append(p_xkb->p_gstring_change_opt_partial, change_opt_id);
+        g_string_append_printf (p_xkb->p_gset_opts, "'%s'", change_opt_id);
 
         //g_printf("\npartial change opt = '%s'\n", p_xkb->p_gstring_change_opt_partial->str);
 
@@ -826,6 +828,7 @@ static void on_button_kbd_change_layout_clicked(GtkButton *p_button, gpointer *p
     if(response == GTK_RESPONSE_OK)
     {
         p_xkb->p_gstring_change_opt_partial = g_string_new("");
+        p_xkb->p_gset_opts = g_string_new ("");
         gtk_tree_model_foreach(GTK_TREE_MODEL(p_liststore_kbd_change),
                                change_opt_tree_model_foreach,
                                p_xkb);
@@ -838,7 +841,17 @@ static void on_button_kbd_change_layout_clicked(GtkButton *p_button, gpointer *p
         config_group_set_string(p_xkb->settings, "ToggleOpt", p_xkb->kbd_change_option);
         g_string_free(p_xkb->p_gstring_change_opt_partial, TRUE/*free also gstring->str*/);
 
+        // update gsettings for mutter
+        char *set = g_strdup_printf ("gsettings set org.gnome.desktop.input-sources xkb-options \"[%s]\"", p_xkb->p_gset_opts->str);
+        system (set);
+        g_free (set);
+        g_string_free (p_xkb->p_gset_opts, TRUE);
+
         gtk_button_set_label(GTK_BUTTON(p_xkb->p_button_change_layout), p_xkb->kbd_change_option);
+#if GTK_CHECK_VERSION(3, 0, 0)
+        gtk_label_set_max_width_chars (GTK_LABEL (gtk_bin_get_child (p_xkb->p_button_change_layout)), 25);
+        gtk_label_set_ellipsize (GTK_LABEL (gtk_bin_get_child (p_xkb->p_button_change_layout)), PANGO_ELLIPSIZE_END);
+#endif
         xkb_setxkbmap(p_xkb);
         xkb_redraw(p_xkb);
     }
@@ -1345,6 +1358,10 @@ static GtkWidget *xkb_configure(LXPanel *panel, GtkWidget *p)
     GtkWidget * p_button_up_layout = gtk_button_new_with_label(_("_Up"));
     GtkWidget * p_button_down_layout = gtk_button_new_with_label(_("_Down"));
     p_xkb->p_button_rm_layout = gtk_button_new_with_label(_("_Remove"));
+    gtk_button_set_use_underline (GTK_BUTTON (p_button_add_layout), TRUE);
+    gtk_button_set_use_underline (GTK_BUTTON (p_button_up_layout), TRUE);
+    gtk_button_set_use_underline (GTK_BUTTON (p_button_down_layout), TRUE);
+    gtk_button_set_use_underline (GTK_BUTTON (p_xkb->p_button_rm_layout), TRUE);
 #else
     GtkWidget * p_button_add_layout = gtk_button_new_from_stock(GTK_STOCK_ADD);
     GtkWidget * p_button_up_layout = gtk_button_new_from_stock(GTK_STOCK_GO_UP);
@@ -1409,6 +1426,8 @@ static GtkWidget *xkb_configure(LXPanel *panel, GtkWidget *p)
     p_xkb->p_button_change_layout = gtk_button_new_with_label(p_xkb->kbd_change_option);
     g_signal_connect(p_xkb->p_button_change_layout, "clicked", G_CALLBACK(on_button_kbd_change_layout_clicked), p_xkb);
 #if GTK_CHECK_VERSION(3, 0, 0)
+    gtk_label_set_max_width_chars (GTK_LABEL (gtk_bin_get_child (p_xkb->p_button_change_layout)), 25);
+    gtk_label_set_ellipsize (GTK_LABEL (gtk_bin_get_child (p_xkb->p_button_change_layout)), PANGO_ELLIPSIZE_END);
     gtk_container_add(GTK_CONTAINER(p_xkb->p_frame_change_layout), p_xkb->p_button_change_layout);
 #else
     gtk_container_add(GTK_CONTAINER(p_alignment_change_layout), p_xkb->p_button_change_layout);
diff --git a/plugins/xkb/xkb.h b/plugins/xkb/xkb.h
index e6ca4f7..cd7c997 100644
--- a/plugins/xkb/xkb.h
+++ b/plugins/xkb/xkb.h
@@ -71,6 +71,7 @@ typedef struct {
     GString  *p_gstring_layouts_partial;
     GString  *p_gstring_variants_partial;
     GString  *p_gstring_change_opt_partial;
+    GString  *p_gset_opts;
     gint      flag_size;
     int       num_layouts;
     gboolean  cust_dir_exists;
